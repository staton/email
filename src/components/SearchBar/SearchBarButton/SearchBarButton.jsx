import React, { Component } from 'react';
import PropTypes from 'prop-types';

const propTypes = {
    content: PropTypes.any.isRequired,
    id: PropTypes.string.isRequired,
    onClick: PropTypes.func,
    title: PropTypes.string
};

const defaultProps = {
    onClick: () => {},
    title: ''
};

export class SearchBarButton extends Component {

    constructor(props) {
        super(props);
    }

	render() {
		return (
            <button 
                className="SearchBarButton"
                id={this.props.id}
                onClick={this.props.onClick}
                onMouseDown={(e) => e.preventDefault()}
                title={this.props.title}
            >
                {this.props.content}
            </button>
		);
    }

}

SearchBarButton.propTypes = propTypes;
SearchBarButton.defaultProps = defaultProps;

export default SearchBarButton;