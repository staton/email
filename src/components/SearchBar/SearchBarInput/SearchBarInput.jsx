import React, { Component } from 'react';
import PropTypes from 'prop-types';
import STRINGS from '../../../resources/strings';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

const propTypes = {
    onBlur: PropTypes.func,
    onChange: PropTypes.func,
    onFocus: PropTypes.func,
    onKeyDown: PropTypes.func,
    placeholder: PropTypes.string
};

const defaultProps = {
    onBlur: () => {},
    onChange: () => {},
    onFocus: () => {},
    onKeyDown: () => {},
    placeholder: STRINGS.SearchPlaceholder
};

export class SearchBarInput extends Component {

    constructor() {
        super();
    }

	render() {
		return (
            <input 
                className="SearchBarInput"
                onBlur={this.props.onBlur}
                onChange={this.props.onChange}
                onFocus={this.props.onFocus}
                onKeyDown={this.props.onKeyDown}
                placeholder={this.props.placeholder}
                type="text"
                value={this.props.searchText} 
            />
		);
    }

}

function mapStateToProps(store, ownProps) {
    return {
        searchText: store.app.searchText,
        onBlur: ownProps.onBlur,
        onChange: ownProps.onChange,
        onFocus: ownProps.onFocus,
        onKeyDown: ownProps.onKeyDown
    };
}

function mapDispatchToProps(dispatch) {
    return bindActionCreators({
    },
    dispatch);
}

SearchBarInput.propTypes = propTypes;
SearchBarInput.defaultProps = defaultProps;

export default connect(mapStateToProps, mapDispatchToProps)(SearchBarInput);