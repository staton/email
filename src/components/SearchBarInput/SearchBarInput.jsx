import React, { Component } from 'react';
import PropTypes from 'prop-types';
import STRINGS from '../../resources/strings';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

const propTypes = {
    onBlur: PropTypes.func,
    onChange: PropTypes.func,
    onFocus: PropTypes.func,
    onKeyPress: PropTypes.func,
    placeholder: PropTypes.string
};

const defaultProps = {
    onBlur: () => {},
    onChange: () => {},
    onFocus: () => {},
    onKeyPress: () => {},
    placeholder: STRINGS.SearchPlaceholder
};

export class SearchBarInput extends Component {

    constructor() {
        super();
    }

	render() {
		return (
            <input 
                className="SearchBarInput"
                onBlur={this.props.onBlur}
                onChange={this.props.onChange}
                onFocus={this.props.onFocus}
                onKeyPress={this.props.onKeyPress}
                placeholder={this.props.placeholder}
                type="text"
                value={this.props.searchText} 
            />
		);
    }

}

function mapStateToProps(store, ownProps) {
    return {
        searchText: store.search.searchText,
        onBlur: ownProps.onBlur,
        onChange: ownProps.onChange,
        onFocus: ownProps.onFocus,
        onKeyPress: ownProps.onKeyPress
    };
}

function mapDispatchToProps(dispatch) {
    return bindActionCreators({
    },
    dispatch);
}

SearchBarInput.propTypes = propTypes;
SearchBarInput.defaultProps = defaultProps;

export default connect(mapStateToProps, mapDispatchToProps)(SearchBarInput);