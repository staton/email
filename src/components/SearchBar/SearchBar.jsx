import React, { Component } from 'react';
import SearchBarButton from './SearchBarButton/SearchBarButton';
import SearchBarInput from './SearchBarInput/SearchBarInput';
import STRINGS from '../../resources/strings';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { MdClear, MdSearch  } from 'react-icons/md';
import { setSearchText } from '../../redux/actions/app-actions';

const unfocusedClassNames = 'search-bar-inner-div';
const focusedClassNames = unfocusedClassNames + ' inner-input-focused';

export class SearchBar extends Component {

    constructor() {
        super();
        
        // In this component, local state will be used to determine the 
        // style of the search bar, based on whether or not it is in focus.
        this.state = { innerDivClassNames: unfocusedClassNames };

        this.handleClearSearchText = this.handleClearSearchText.bind(this);
        this.handleSearch = this.handleSearch.bind(this);
        this.handleSearchBlurred = this.handleSearchBlurred.bind(this);
        this.handleSearchFocused = this.handleSearchFocused.bind(this);
        this.handleSearchKeyDown = this.handleSearchKeyDown.bind(this);
        this.handleSearchTextChanged = this.handleSearchTextChanged.bind(this);
    }

	render() {
        var innerDivClassNames = this.state.innerDivClassNames;

		return (
            <div 
                className="SearchBar"
                style={{ backgroundColor: this._backgroundColor }}
            >
                <div className={innerDivClassNames}>
                    <SearchBarButton 
                        content={<MdSearch />}
                        id="search-bar-search-button"
                        onClick={this.search}
                        title={STRINGS.Search}
                    />
                    <SearchBarInput 
                        onBlur={this.handleSearchBlurred}
                        onChange={this.handleSearchTextChanged}
                        onFocus={this.handleSearchFocused}
                        onKeyDown={this.handleSearchKeyDown}
                    />
                    {
                        // only display the clear-search button if the search bar contains text
                        (this.props.searchText)
                        ?   <SearchBarButton
                                content={<MdClear />}
                                id="search-bar-clear-button"
                                onClick={this.handleClearSearchText}
                                title={STRINGS.Clear}
                            />
                        :   null
                    }
                </div>
            </div>
		);
    }
    
    /**
     * Clears the current search bar input text.
     * @param {object} e The event object.
     */
    handleClearSearchText(e) {
        if (this.props.searchText) {
            this.props.setSearchText('');
        }
    }

    /**
     * Performs a search based on the current search bar input text.
     * @param {object} e The event object.
     */
    handleSearch(e) {
        if (this.props.searchText) {

        }
    }

    /**
     * Called when the search input loses focus.
     * @param {object} e The event object.
     */
    handleSearchBlurred(e) {
        // remove the focused class name from this element, to revert to the
        // search bar's default style.
        this.setState({ innerDivClassNames: unfocusedClassNames });
    }

    /**
     * Called when the search input receives focus.
     * @param {object} e The event object.
     */
    handleSearchFocused(e) {
        // add the focused class name to this element, which will control
        // the search bar style when it is focused.
        this.setState({ innerDivClassNames: focusedClassNames });
    }

    /**
     * Called when there is a key-down event in the search input.
     * @param {object} e The event object.
     */
    handleSearchKeyDown(e) {
        
    }

    /**
     * Called when the search input changes.
     * @param {object} e The event object.
     */
    handleSearchTextChanged(e) {
        if (e.which === 13) {
            console.log('enter pressed!!!!!!');
        } else if (this.props.searchText !== e.target.value) {
            this.props.setSearchText(e.target.value);
        }
    }

}

function mapStateToProps(store, ownProps) {
    return {
        searchText: store.app.searchText
    };
}

function mapDispatchToProps(dispatch) {
    return bindActionCreators({
        setSearchText: setSearchText
    },
    dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(SearchBar);