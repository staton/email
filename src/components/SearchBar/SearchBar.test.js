import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import { SearchBar } from './SearchBar';
import SearchBarButton from './SearchBarButton/SearchBarButton';
import SearchBarInput from './SearchBarInput/SearchBarInput';

describe('<SearchBar />', () => {

    it('renders correctly with no text', () => {
        const wrapper = shallow(<SearchBar searchText="" />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders correctly with text', () => {
        const wrapper = shallow(<SearchBar searchText="hello" />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    
    it('contains an input and search button when input contains no text', () => {
        const wrapper = shallow(<SearchBar searchText="" />);
        expect(wrapper.find(SearchBarInput).length).toEqual(1);
        expect(wrapper.find(SearchBarButton).length).toEqual(1);
        //expect(wrapper.find('#search-bar-search-button').length).toEqual(1);
        //expect(wrapper.find('#search-bar-clear-button').length).toEqual(0);
    });

    it('contains an input, search button, and clear button when input contains text', () => {
        const wrapper = shallow(<SearchBar searchText="hello" />);
        expect(wrapper.find(SearchBarInput).length).toEqual(1);
        expect(wrapper.find(SearchBarButton).length).toEqual(2);
        //expect(wrapper.find('#search-bar-search-button').length).toEqual(1);
        //expect(wrapper.find('#search-bar-clear-button').length).toEqual(1);
    });

});