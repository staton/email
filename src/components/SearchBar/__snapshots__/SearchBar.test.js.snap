// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> renders correctly with no text 1`] = `
<div
  className="SearchBar"
  style={
    Object {
      "backgroundColor": undefined,
    }
  }
>
  <div
    className="search-bar-inner-div"
  >
    <SearchBarButton
      content={<Unknown />}
      id="search-bar-search-button"
      onClick={[Function]}
      title="Search"
    />
    <Connect(SearchBarInput)
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
    />
  </div>
</div>
`;

exports[`<SearchBar /> renders correctly with text 1`] = `
<div
  className="SearchBar"
  style={
    Object {
      "backgroundColor": undefined,
    }
  }
>
  <div
    className="search-bar-inner-div"
  >
    <SearchBarButton
      content={<Unknown />}
      id="search-bar-search-button"
      onClick={[Function]}
      title="Search"
    />
    <Connect(SearchBarInput)
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
    />
    <SearchBarButton
      content={<Unknown />}
      id="search-bar-clear-button"
      onClick={[Function]}
      title="Clear"
    />
  </div>
</div>
`;
